src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp

```diff
@@ -884,25 +888,28 @@ bool CachedOnDiskReadBufferFromFile::nextImplStep()
 
     if (!result)
     {
-#ifndef NDEBUG
-        if (read_type == ReadType::CACHED)
+        auto debug_check = [&]()
         {
-            size_t cache_file_size = getFileSizeFromReadBuffer(*implementation_buffer);
-            if (cache_file_size == 0)
+            if (read_type == ReadType::CACHED)
             {
-                throw Exception(
-                    ErrorCodes::LOGICAL_ERROR,
-                    "Attempt to read from an empty cache file: {} (just before actual read)",
-                    cache_file_size);
+                size_t cache_file_size = getFileSizeFromReadBuffer(*implementation_buffer);
+                if (cache_file_size == 0)
+                {
+                    throw Exception(
+                        ErrorCodes::LOGICAL_ERROR,
+                        "Attempt to read from an empty cache file: {} (just before actual read)",
+                        cache_file_size);
+                }
             }
-        }
-        else
-        {
-            assert(file_offset_of_buffer_end == static_cast<size_t>(implementation_buffer->getFileOffsetOfBufferEnd()));
-        }
+            else
+            {
+                chassert(file_offset_of_buffer_end == static_cast<size_t>(implementation_buffer->getFileOffsetOfBufferEnd()));
+            }
+            chassert(!implementation_buffer->hasPendingData());
+            return true;
+        };
 
-        assert(!implementation_buffer->hasPendingData());
-#endif
+        chassert(debug_check());
```

Comment:
it's not clear to me why we need to call this block through `chassert` - `debug_check` always returns `true`.

Reply:
the point is `debug_check()` throws exception or fails chassert's inside it, when something is wrong, therefore this lambda never returns false. The point of putting it in chassert - to make this debug check be executed not only in debug build, but also in builds with sanitisers, while previously it was under `#ifndef NDEBUG` and was executed only in debug build. So it does not matter that `debug_check` always returns `true` as otherwise exception or abort from chassert inside the check is expected.

Reply:
because it is better not to do `stat` all the time in release build (to get file size)

Reply:
oh, I see - missed that `chassert` is defined only with `ABORT_ON_LOGICAL_ERROR`